cli/ts/commands/joinPoll.ts

```diff
@@ -0,0 +1,257 @@
+import { extractVk, genProof, verifyProof } from "maci-circuits";
+import { formatProofForVerifierContract, genMaciStateFromContract } from "maci-contracts";
+import { MACI__factory as MACIFactory, Poll__factory as PollFactory } from "maci-contracts/typechain-types";
+import { CircuitInputs, IJsonMaciState, MaciState } from "maci-core";
+import { poseidon } from "maci-crypto/build/ts/hashing";
+import { Keypair, PrivKey, PubKey } from "maci-domainobjs";
+
+import fs from "fs";
+
+import type { IJoinPollArgs, IJoinedUserArgs, IParsePollJoinEventsArgs } from "../utils/interfaces";
+
+import { contractExists, logError, logYellow, info, logGreen, success } from "../utils";
+import { banner } from "../utils/banner";
+
+export const joinPoll = async ({
+  maciAddress,
+  privateKey,
+  pollPrivKey,
+  stateIndex,
+  newVoiceCreditBalance,
+  stateFile,
+  pollId,
+  signer,
+  startBlock,
+  endBlock,
+  blocksPerBatch,
+  transactionHash,
+  pollJoiningZkey,
+  useWasm,
+  rapidsnark,
+  pollWitgen,
+  pollWasm,
+  quiet = true,
+}: IJoinPollArgs): Promise<number> => {
+  banner(quiet);
+  const userSideOnly = true;
+
+  if (!(await contractExists(signer.provider!, maciAddress))) {
+    logError("MACI contract does not exist");
+  }
+
+  if (!PrivKey.isValidSerializedPrivKey(privateKey)) {
+    logError("Invalid MACI private key");
+  }
+
+  const userMaciPrivKey = PrivKey.deserialize(privateKey);
+  const nullifier = poseidon([BigInt(userMaciPrivKey.asCircuitInputs())]);
+
+  // Create poll public key from poll private key
+  const pollPrivKeyDeserialized = PrivKey.deserialize(pollPrivKey);
+  const pollKeyPair = new Keypair(pollPrivKeyDeserialized);
+  const pollPubKey = pollKeyPair.pubKey;
+
+  // check < 1 cause index zero is a blank state leaf
+  if (stateIndex < 1) {
+    logError("Invalid state index");
+  }
+
+  if (pollId < 0) {
+    logError("Invalid poll id");
+  }
+
+  const maciContract = MACIFactory.connect(maciAddress, signer);
+  const pollAddress = await maciContract.getPoll(pollId);
+
+  if (!(await contractExists(signer.provider!, pollAddress))) {
+    logError("Poll contract does not exist");
+  }
+
+  const pollContract = PollFactory.connect(pollAddress, signer);
+
+  let maciState: MaciState | undefined;
+  if (stateFile) {
+    const content = JSON.parse(fs.readFileSync(stateFile).toString()) as unknown as IJsonMaciState;
```

Comment:
Use async version for reading files instead

Reply:
Done